controller.py

```diff
@@ -138,3 +148,50 @@ def search(self, ip_address=None):
                 3600 * 24 * 30
             )
             return Response(body, 200, headers)
+
+    def register(self, do_get=HTTP.get_with_timeout):
+        opds_url = flask.request.form.get("url")
+        if not opds_url:
+            return NO_OPDS_URL
+
+        try:
+            response = do_get(opds_url, allowed_response_codes=["2xx", "3xx"])
+            feed = feedparser.parse(response.content)
+        except Exception, e:
+            return INVALID_OPDS_FEED
+        links = feed.get("feed", {}).get("links", [])
+
+        auth_url = None
+        for link in links:
+            # TODO: Define a rel for the auth document itself, and
+            # look for that instead.
+            if link.get("rel") == "http://opds-spec.org/shelf":
+                auth_url = link.get("href")
+                break
+
+        if not auth_url:
+            return NO_SHELF_LINK
+
+        try:
+            response = do_get(auth_url, allowed_response_codes=["2xx", "3xx", "4xx"])
+            auth_document = json.loads(response.content)
+        except Exception, e:
+            return INVALID_AUTH_DOCUMENT
+
+        library, is_new = get_one_or_create(
+            self._db, Library,
+            opds_url=opds_url
+        )
+
+        library.name = auth_document.get("name")
```

Comment:
It might be premature, but given the number of extensions we've added to A4OPDS I think it's a good idea to create a data object that can parse an A4OPDS document. If nothing else it'll make it a lot easier to write tests that handle various weird documents.